### PR TITLE
feat: Add multi-CUDA version support (12.8, 12.9, 13.0, 13.1)

### LINKED ISSUE #4: Keep the ODH cuda images close to what Nvidia publishes
~~As suggested by @ktdreyer https://github.com/opendatahub-io/base-containers/pull/1/changes#r2695498454 this looks expensive to maintain. Do we really need it, if yes then we need an easy way to maintain this~~

~~NVIDIA_REQUIRE_CUDA="cuda>=12.8 brand=unknown,driver>=470,driver<471 brand=grid,driver>=470,driver<471 brand=tesla,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=vapps,driver>=470,driver<471 brand=vpc,driver>=470,xxx~~

The goal has changed now to maintain multiple CUDA versions, so the original issue is not valid as we do not have update the cantainerfiles for newer version, but we will create new containerfiles.

### PR DESCRIPTION
**Pull request by LalatenduMohanty**
Refactor to version-based directory structure following NVIDIA's pattern [1] .Each version has its own Containerfile and app.conf.

- Add templates and generate script for new versions
- Update build.sh with versioned targets (cuda-12.8, python-3.12)
- Update CI with matrix builds for all versions

[1] https://gitlab.com/nvidia/container-images/cuda

Closes #4

## Summary by CodeRabbit

* **New Features**
  * Per-version Python (3.12) and CUDA (12.8, 12.9, 13.0, 13.1) images with per-version build/test matrix and selectable build targets
  * Added template-driven generation tooling to create version-specific image definitions

* **Documentation**
  * Updated docs and quick-start for per-version layout and how to add new versions

* **Tests**
  * Tests read version env vars, validate conditionally, and skip when versions are unset

* **Chores**
  * Repo reorganized to per-version structure; CI and lint flows updated; build/dist ignored in repo rules

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -27,16 +32,51 @@ jobs:
...
+              - 'cuda/**'
+              - 'Containerfile.cuda.template'
+              - 'scripts/build.sh'
+              - 'tests/test_cuda_image.py'
+              - 'tests/test_common.py'
+              - 'tests/conftest.py'
+            any-python:
+              - 'python/**'
+              - 'Containerfile.python.template'
               - 'scripts/build.sh'
```
Comment (shifa-khan) [inline .github/workflows/ci.yml]: Should we consider adding CI check to verify the generated Containerfiles stay in sync with templates?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -147,28 +191,37 @@ jobs:
...
     if: |
       always() &&
-      needs.changes.outputs.cuda == 'true' &&
+      needs.changes.outputs['any-cuda'] == 'true' &&
       needs.lint.result == 'success' &&
       needs.type-check.result == 'success' &&
       (needs.lint-containerfiles.result == 'success' || needs.lint-containerfiles.result == 'skipped')
+    strategy:
+      matrix:
+        version: ["12.8", "12.9", "13.0", "13.1"]
```
[the remark this answers is not in the document]
Comment (LalatenduMohanty) [reply .github/workflows/ci.yml]: You are right. However 

* It will add ~20 lines of matrix generation logic
* Has edge cases (template changes, test changes)
* GitHub Actions dynamic matrices can be tricky

This is  a valid optimization but adds complexity. What you think about tracking this in a separate issue and fix it in a follow-up PR.

The current approach works correctly, just slower

File: tests/test_cuda_image.py [hunk 3]
```diff
@@ -4,13 +4,22 @@
...
 
 def test_cuda_version(cuda_container):
-    """Verify CUDA_VERSION environment variable is set to 12.8.x."""
-    assert cuda_container.get_env("CUDA_VERSION").startswith("12.8")
+    """Verify CUDA_VERSION environment variable matches expected version.
+
+    The expected version is controlled by the CUDA_VERSION environment variable.
+    If not set, defaults to 12.8 for backward compatibility.
+    """
+    expected_version = os.environ.get("CUDA_VERSION", "12.8")
```
[the remark this answers is not in the document]
Comment (LalatenduMohanty) [reply tests/test_cuda_image.py]: Fixed it.

**Review by smoparth — APPROVED** [review]
LGTM

File: .tekton/odh-midstream-cuda-base-main-pull-request.yaml [hunk 4]
```diff
@@ -31,9 +31,9 @@ spec:
     value:
     - linux/x86_64
   - name: dockerfile
-    value: Containerfile.cuda
+    value: cuda/12.8/Containerfile
   - name: build-args-file
-    value: build-args/cuda-app.conf
+    value: cuda/12.8/app.conf
```
[the remark this answers is not in the document]
Comment (LalatenduMohanty) [reply .tekton/odh-midstream-cuda-base-main-pull-request.yaml]: @Victoremepunto Thanks for the clarification. I will go ahead and merge this PR. Once you take a look create follow up issues if required for this change. We will fix them.

**Comment by LalatenduMohanty** [comment]
Because of https://us.githubstatus.com/incidents/4crg7jymnmf7  the pre-submit jobs are getting skipped
```
We continue to investigate failures impacting GitHub Actions hosted-runner jobs.
We have identified the root cause and are working with our upstream provider to mitigate.
This is also impacting GitHub features that rely on GitHub Actions (for example, Copilot Coding Agent and Dependabot).
```
